lib/AST/DeclContext.cpp

```diff
@@ -1035,6 +1035,25 @@ DeclContextKind DeclContext::getContextKind() const {
   llvm_unreachable("Unhandled DeclContext ASTHierarchy");
 }
 
+bool DeclContext::hasValueSemantics() const {
+  if (auto contextTy = getSelfTypeInContext()) {
```

Comment:
This can trigger a lot of work, for example computing the generic signature, even in cases that don't need to know the generic signature (ie, anything other than a protocol extension). Maybe it's best to handle the protocol extension case separately, like so:

```
if (getExtendedProtocolDecl())
  return isClassConstrainedProtocolExtension();
return getDeclaredType()->hasReferenceSemantics();
```